clippy_lints/src/iter_without_into_iter.rs

```diff
@@ -0,0 +1,149 @@
+    /// (`for val in &iter {}`), without having to first call `iter()` or `iter_mut()`.
+    ///
+    /// ### Example
+    /// ```rust
+    /// struct MySlice<'a>(&'a [u8]);
+    /// impl<'a> MySlice<'a> {
+    ///     pub fn iter(&self) -> std::slice::Iter<'a, u8> {
+    ///         self.0.iter()
+    ///     }
+    /// }
+    /// ```
+    /// Use instead:
+    /// ```rust
+    /// struct MySlice<'a>(&'a [u8]);
+    /// impl<'a> MySlice<'a> {
+    ///     pub fn iter(&self) -> std::slice::Iter<'a, u8> {
+    ///         self.0.iter()
+    ///     }
+    /// }
+    /// impl<'a> IntoIterator for &MySlice<'a> {
+    ///     type Item = &'a u8;
+    ///     type IntoIter = std::slice::Iter<'a, u8>;
+    ///     fn into_iter(self) -> Self::IntoIter {
+    ///         self.iter()
+    ///     }
+    /// }
+    /// ```
+    #[clippy::version = "1.74.0"]
+    pub ITER_WITHOUT_INTO_ITER,
+    pedantic,
+    "implementing `iter(_mut)` without an associated `IntoIterator for (&|&mut) Type` impl"
+}
+declare_lint_pass!(IterWithoutIntoIter => [ITER_WITHOUT_INTO_ITER]);
+
+/// Checks if a given type is nameable in a trait (impl).
+/// RPIT is stable, but impl Trait in traits is not (yet), so when we have
+/// a function such as `fn iter(&self) -> impl IntoIterator`, we can't
+/// suggest `type IntoIter = impl IntoIterator`.
+fn is_nameable_in_impl_trait(ty: &rustc_hir::Ty<'_>) -> bool {
+    !matches!(ty.kind, TyKind::OpaqueDef(..))
+}
+
+fn check_for_mutability(cx: &LateContext<'_>, item: &rustc_hir::ImplItem<'_>, expected_mtbl: Mutability) {
+    let (borrow_prefix, name, expected_implicit_self) = match expected_mtbl {
+        Mutability::Not => ("&", "iter", ImplicitSelfKind::ImmRef),
+        Mutability::Mut => ("&mut ", "iter_mut", ImplicitSelfKind::MutRef),
+    };
+
+    if let ImplItemKind::Fn(sig, _) = item.kind
+        && let FnRetTy::Return(ret) = sig.decl.output
+        && is_nameable_in_impl_trait(ret)
+        && cx.tcx.generics_of(item.owner_id.def_id).params.is_empty()
+        && sig.decl.implicit_self == expected_implicit_self
+        && sig.decl.inputs.len() == 1
+        && let Some(imp) = get_parent_as_impl(cx.tcx, item.hir_id())
+        && imp.of_trait.is_none()
+        && let TyKind::Path(QPath::Resolved(_, path)) = imp.self_ty.kind
+        && let Some(ty_did) = path.res.opt_def_id()
+        && let middle_ty = cx.tcx.type_of(ty_did).skip_binder()
+        && let ref_ty = cx.tcx.mk_ty_from_kind(ty::Ref(cx.tcx.lifetimes.re_erased, middle_ty, expected_mtbl))
+        && let Some(into_iter_did) = cx.tcx.get_diagnostic_item(sym::IntoIterator)
+        && let bound_vars = cx.tcx.late_bound_vars(item.hir_id())
+        // We need to erase late bounds regions so that `-> Iter<'_, T>` works
+        && let ret_middle_ty = cx.tcx.erase_late_bound_regions(
+            ty::Binder::bind_with_vars(hir_ty_to_ty(cx.tcx, ret), bound_vars)
+        )
```

Comment:
I'm not 100% confident if this is right. `implements_trait` just returns false if we're passing it a type like `Iter<'_, T>`, probably because of this:
https://github.com/rust-lang/rust-clippy/blob/251a475b724a81ffcfd0113820492a329feeef5c/clippy_utils/src/ty.rs#L242-L243

And erasing the regions seems to make it work. Is it right to use the bound vars from the fn?

Reply:
You'll want to use `liberate_late_bound_regions` on the functions signature and get the types from there.